helix-core/src/comment.rs

```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
     Transaction::change(doc, changes.into_iter())
 }
 
+fn find_last_non_whitespace_char(text: RopeSlice) -> Option<usize> {
+    (0..text.len_chars()).rev().find(|&i| {
+        text.get_char(i)
+            .map(|ch| !ch.is_whitespace())
+            .unwrap_or(false)
+    })
+}
+
+#[derive(Debug, PartialEq)]
+struct CommentChange {
+    range: Range,
+    open_pos: usize,
+    close_pos: usize,
+    open_margin: usize,
+    close_margin: usize,
+}
+
+fn find_block_comments(
+    open: &str,
+    close: &str,
+    text: RopeSlice,
+    selection: &Selection,
+) -> (bool, Vec<CommentChange>) {
+    let mut commented = true;
+    let mut comment_changes = Vec::with_capacity(selection.len());
+    for range in selection {
+        let selection_slice = range.slice(text);
+        if let (Some(open_pos), Some(close_pos)) = (
+            find_first_non_whitespace_char(selection_slice),
+            find_last_non_whitespace_char(selection_slice),
+        ) {
+            let len = selection_slice.len_chars();
+
+            // selection can be shorter than open_pos + open len
+            let open_fragment = Cow::from(
+                selection_slice.slice(open_pos..std::cmp::min(open_pos + open.len(), len)),
+            );
+            let close_fragment = Cow::from(
+                selection_slice
+                    .slice(close_pos - std::cmp::min(close.len() - 1, close_pos)..close_pos + 1),
+            );
+
+            // margin of one if there is a space between the comment token and other characters
+            let open_margin = match selection_slice.get_char(open_pos + open.len()) {
+                Some(c) if c == ' ' => 1,
+                _ => 0,
+            };
+
+            let close_margin = if open_pos + open.len()
+                != close_pos - std::cmp::min(close.len(), close_pos)
+            {
+                match selection_slice.get_char(close_pos - std::cmp::min(close.len(), close_pos)) {
+                    Some(c) if c == ' ' => 1,
+                    _ => 0,
+                }
+            } else {
+                0
+            };
```

Comment:
this code is hard to read. I would simply do this after you have determined that there is indeed a block comment. Then you can calculate the inner slice and use `get_char(0)` and `get_char(slice.len().satusating_sub(1))`. 

you can use `map_or` to make this more readable, also store the margins as a bools. You can then use `as usize` when generating the actual transaction